File: ts/src/day03.ts
import { readFileSync, exists } from "fs";

interface Rectangle {
    left: number
    top: number
    width: number
    height: number
}

interface Claim extends Rectangle {
    id: number
}

function parseClaim(text: string): Claim | undefined {
    let match = text.match(/#(\d+) @ (\d+),(\d+): (\d+)x(\d+)/);
    if (match) {
        let r = {
            id: Number(match[1]),
            left: Number(match[2]),
            top: Number(match[3]),
            width: Number(match[4]),
            height: Number(match[5]),
        }
        return r;
    }
}

function calculateOverlap(r1: Rectangle, r2: Rectangle): Rectangle | undefined {
    let left = Math.max(r1.left, r2.left);
    let right = Math.min(r1.left + r1.width, r2.left + r2.width);
    let top = Math.max(r1.top, r2.top);
    let bottom = Math.min(r1.top + r1.height, r2.top + r2.height);
    if (left >= right || top >= bottom) {
        return;
    }
    return {
        left,
        top,
        width: right - left,
        height: bottom - top,
    };
}

function containsPoint(r: Rectangle) {
    return (x: number, y: number) => {
        return (
            (r.left <= x && x < (r.left + r.width)) && 
            (r.top <= y && y < (r.top + r.height))
        )
    }
}

function anyContainsPoint(rs: Rectangle[]) {
    let contains = rs.map(containsPoint);
    return (x: number, y: number) => contains.some(c => c(x, y));
}

function calculateBounds(rs: Rectangle[]): Rectangle {
    let left = Number.MAX_SAFE_INTEGER;
    let right = 0;
    let top = Number.MAX_SAFE_INTEGER;
    let bottom = 0;

    for (let r of rs) {
        left = Math.min(left, r.left);
        right = Math.max(right, r.left + r.width);
        top = Math.min(top, r.top);
        bottom = Math.max(bottom, r.top + r.height);
    }

    return {
        left,
        top,
        width: right - left,
        height: bottom - top,
    }
}

function getInput(): Claim[] {
    const input = readFileSync("./src/inputs/day03.txt", "utf8");
    return input
        .split("\n")
        .filter(line => line.length > 0)
        .map(parseClaim)
        .filter(claim => claim !== undefined) as Claim[];
}

export function part1() {
    let claims = getInput();
    let overlaps = [];
    for (let i = 0; i < claims.length; i++) {
        for (let j = i + 1; j < claims.length; j++) {
            let overlap = calculateOverlap(claims[i], claims[j]);
            if (overlap !== undefined) {
                overlaps.push(overlap);
            }
        }
    }
    console.log(`Found ${overlaps.length} overlaps`);
    let boundingRectangle = calculateBounds(overlaps);
    console.log(`Bounding rectangle:`, boundingRectangle);
    let contains = anyContainsPoint(overlaps);
    let area = 0;
    for (let i = boundingRectangle.left; i < boundingRectangle.left + boundingRectangle.width; i++) {
        for (let j = boundingRectangle.top; j < boundingRectangle.top + boundingRectangle.height; j++) {
            area += contains(i, j) ? 1 : 0;
        }
    }
    return area;
}

export function part2() {
    let claims = getInput();
    let allIds = claims.map(c => c.id);
    for (let i = 0; i < claims.length; i++) {
        for (let j = i + 1; j < claims.length; j++) {
            let overlap = calculateOverlap(claims[i], claims[j]);
            if (overlap !== undefined) {
                let id1 = claims[i].id;
                let id2 = claims[j].id;
                allIds = allIds.filter(id => id !== id1).filter(id => id !== id2);
            }
        }
    }
    return allIds[0];
}

File: rs/src/day03.rs
use regex::Regex;

use std::collections::HashSet;

const INPUT: &'static str = include_str!("../inputs/day03.txt");

struct Rect {
    pub left: i32,
    pub top: i32,
    pub width: i32,
    pub height: i32,
}

struct Claim {
    pub id: i32,
    pub r: Rect,
}

impl Rect {
    pub fn new(left: i32, top: i32, width: i32, height: i32) -> Rect {
        Rect {
            left: left,
            top: top,
            width: width,
            height: height,
        }
    }

    pub fn right(&self) -> i32 {
        self.left + self.width
    }

    pub fn bottom(&self) -> i32 {
        self.top + self.height
    }

    pub fn overlap(&self, other: &Rect) -> Option<Rect> {
        let left = std::cmp::max(self.left, other.left);
        let right = std::cmp::min(self.right(), other.right());
        let top = std::cmp::max(self.top, other.top);
        let bottom = std::cmp::min(self.bottom(), other.bottom());

        if left < right && top < bottom {
            Some(Rect::new(left, top, right - left, bottom - top))
        } else {
            None
        }
    }

    pub fn contains(&self, x: i32, y: i32) -> bool {
        (self.left <= x) && (x < self.right()) && (self.top <= y) && (y < self.bottom())
    }
}

fn any_contains(rects: &Vec<Rect>, x: i32, y: i32) -> bool {
    rects.iter().any(|ref r| r.contains(x, y))
}

impl Claim {
    pub fn new(id: i32, rect: Rect) -> Claim {
        Claim { id: id, r: rect }
    }

    pub fn parse(input: &str) -> Option<Claim> {
        let re = Regex::new("(\\d+)").unwrap();

        let matches: Vec<&str> = re.find_iter(input).map(|m| m.as_str()).collect();

        if matches.len() != 5 {
            return None;
        }

        let id = matches.get(0)?.parse().ok()?;
        let left = matches.get(1)?.parse().ok()?;
        let top = matches.get(2)?.parse().ok()?;
        let width = matches.get(3)?.parse().ok()?;
        let height = matches.get(4)?.parse().ok()?;

        Some(Claim::new(id, Rect::new(left, top, width, height)))
    }
}

fn get_input() -> Vec<Claim> {
    let input = INPUT.to_string();

    input
        .split("\n")
        .into_iter()
        .map(Claim::parse)
        .filter(|o| o.is_some())
        .map(|o| o.unwrap())
        .collect()
}


fn part1() -> i32 {
    let claims = get_input();
    let mut overlaps: Vec<Rect> = Vec::new();

    for i in 0..claims.len() {
        for j in (i + 1)..claims.len() {
            let overlap = claims[i].r.overlap(&claims[j].r);
            if overlap.is_some() {
                overlaps.push(overlap.unwrap())
            }
        }
    }

    let mut area = 0i32;

    for i in 0..1000 {
        for j in 0..1000 {
            area += if any_contains(&overlaps, i, j) { 1 } else { 0 }
        }
    }

    area
}

fn part2() -> i32 {
    let claims = get_input();
    let all_ids: HashSet<i32> = claims.iter().map(|c| c.id).collect();
    let mut overlapping_ids: HashSet<i32> = HashSet::new();
    for i in 0..claims.len() {
        for j in (i + 1)..claims.len() {
            let overlap = claims[i].r.overlap(&claims[j].r);
            if overlap.is_some() {
                overlapping_ids.insert(claims[i].id);
                overlapping_ids.insert(claims[j].id);
            }
        }
    }

    let missing = all_ids.difference(&overlapping_ids);
    *missing.last().unwrap_or(&0)
}

pub fn run() {
    println!("{}", part1());
    println!("{}", part2());
}
